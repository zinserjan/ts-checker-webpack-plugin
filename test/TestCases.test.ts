import * as path from "path";
import * as fs from "fs-extra";
import webpack = require("webpack");
import MemoryFs = require("memory-fs");
import pDefer = require("p-defer");
import { createAssertExpectation, satisfiesVersionRequirements } from "./_util/testHelper";

const testCasesPath = path.join(__dirname, "testCases");
const tests = fs.readdirSync(testCasesPath).filter(dir => fs.statSync(path.join(testCasesPath, dir)).isDirectory());

const tmpPath = path.resolve(__dirname, "../.tmp/testCases");

jasmine.DEFAULT_TIMEOUT_INTERVAL = 15000;

describe("TestCases", () => {
  beforeAll(() => {
    return fs.emptyDir(tmpPath);
  });

  tests.forEach(testName => {
    const testPath = path.join(testCasesPath, testName);
    const tmpTestPath = path.join(tmpPath, testName);
    const webpackConfigPath = path.join(tmpTestPath, "webpack.config.ts");
    const skipTest = !satisfiesVersionRequirements(path.join(testPath, "versions.json"));
    const assertExpectation = createAssertExpectation(path.join(testPath, "expectation.ts"));

    (skipTest ? it.skip : it)(testName, async () => {
      await fs.copy(testPath, tmpTestPath);

      const webpackConfig = require(webpackConfigPath);
      const c = webpack(webpackConfig);
      const compilers = (c as any).compilers ? (c as any).compilers : [c];

      compilers.forEach((compiler: any) => {
        const ifs = compiler.inputFileSystem;
        compiler.inputFileSystem = Object.create(ifs);
        compiler.outputFileSystem = new MemoryFs();
        compiler.inputFileSystem.readFile = function() {
          const args = Array.prototype.slice.call(arguments);
          const callback = args.pop();
          ifs.readFile.apply(
            ifs,
            args.concat([
              (err: any, result: any) => {
                if (err) return callback(err);
                callback(null, result.toString("utf-8").replace(/\r/g, ""));
              },
            ])
          );
        };
        compiler.apply(new webpack.optimize.OccurrenceOrderPlugin(false));
      });

      const deferred = pDefer();
      c.run((err: Error, stats: webpack.Stats) => {
        if (err) {
          return deferred.reject(err);
        }

        assertExpectation(stats, 0);

        process.nextTick(deferred.resolve);
      });
      return deferred.promise;
    });
  });
});
