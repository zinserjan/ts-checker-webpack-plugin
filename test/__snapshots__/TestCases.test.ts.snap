// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestCases basic-type-error 1`] = `
Object {
  "errors": Array [
    "xfile
(1,7): error TS2322",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases basic-type-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`TestCases code-splitting-import-error 1`] = `
Object {
  "errors": Array [
    "xfile
(1,1): error TS2304",
    "xfile
(1,7): error TS2322",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases code-splitting-import-no-error 1`] = `
Object {
  "errors": Array [
    "xfile
(1,1): error TS2304",
    "xfile
(4,1): error TS2304",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases code-splitting-import-ts-dependency-error 1`] = `
Object {
  "errors": Array [
    "xfile
(1,1): error TS2304",
    "xfile
(1,7): error TS2322",
    "xfile
(3,22): error TS2420",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases code-splitting-import-ts-dependency-no-error 1`] = `
Object {
  "errors": Array [
    "xfile
(1,1): error TS2304",
    "xfile
(4,1): error TS2304",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases commonjs-usage 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`TestCases css-modules-error 1`] = `
Object {
  "errors": Array [
    "xfile
(3,20): error TS2339",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases css-modules-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`TestCases ignore-child-compiler-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`TestCases skip-type-check-on-build-error 1`] = `
Object {
  "errors": Array [
    "xfile
(3,1): error TS1005",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases ts-dependency-only-error 1`] = `
Object {
  "errors": Array [
    "xfile
(3,22): error TS2420",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases ts-dependency-only-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;

exports[`TestCases tsconfig-invalid-option 1`] = `
Object {
  "errors": Array [
    "xfile
error TS5023: Unknown compiler option 'xxxx'.",
    "xfile
Module build failed: error while parsing tsconfig.json",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases tsconfig-syntax-error 1`] = `
Object {
  "errors": Array [
    "xfile
Module build failed: error TS5014: Failed to parse file 'xfile': Unexpected token } in JSON at position 218.",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases tslint-error 1`] = `
Object {
  "errors": Array [
    "xfile
  2:11  error  ' should be \\"  quotemark
",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases tslint-warning 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [
    "xfile
  2:11  warning  ' should be \\"  quotemark
",
  ],
}
`;

exports[`TestCases use-library-type-definition-error 1`] = `
Object {
  "errors": Array [
    "xfile
(1,24): error TS7016",
  ],
  "warnings": Array [],
}
`;

exports[`TestCases use-library-type-definition-no-error 1`] = `
Object {
  "errors": Array [],
  "warnings": Array [],
}
`;
